q2_amr/card/partition.py

```diff
@@ -98,3 +101,115 @@ def _collate(partition_list):
+
+
+def _partition_annotations(
+    annotations: Union[
+        CARDAnnotationDirectoryFormat,
+        CARDGeneAnnotationDirectoryFormat,
+        CARDAlleleAnnotationDirectoryFormat,
+    ],
+    num_partitions: int = None,
+):
+    partitioned_annotations = {}
+    annotations_all = []
+    # Add one tuples with sample ID, MAG ID and full paths to annotation files to
+    # annotations_all
+    if isinstance(annotations, CARDAnnotationDirectoryFormat):
+        for sample_id, mag in annotations.sample_dict().items():
+            for mag_id, file_paths in mag.items():
+                annotations_all.append((sample_id, mag_id, file_paths))
+
+    else:
+        for sample_id, file_paths in annotations.sample_dict().items():
+            annotations_all.append((sample_id, file_paths))
+
+    # Sort annotations_all for consistent splitting behaviour
+    annotations_all.sort()
+
+    # Retrieve the number of annotations
+    num_annotations = len(annotations_all)
+
+    # If no number of partitions is specified or the number is higher than the number
+    # of annotations, all annotations get partitioned by annotation
+    if num_partitions is None:
+        num_partitions = num_annotations
+    elif num_partitions > num_annotations:
+        warnings.warn(
+            "You have requested a number of partitions"
+            f" '{num_partitions}' that is greater than your number"
+            f" of annotations '{num_annotations}'. Your data will be"
+            f" partitioned by annotation into '{num_annotations}'"
+            " partitions."
+        )
+        num_partitions = num_annotations
+
+    # Splits annotations into the specified number of arrays
+    arrays = np.array_split(np.array(annotations_all, dtype=object), num_partitions)
+
+    for i, annotation_tuple in enumerate(arrays, 1):
+        # Creates directory with same format as input
+        partitioned_annotation = type(annotations)()
+
+        # Constructs paths to all annotation files and moves them to the new partition
+        # directories
+        if isinstance(annotations, CARDAnnotationDirectoryFormat):
+            for sample_id, mag_id, file_paths in annotation_tuple:
+                copy_files(file_paths, partitioned_annotation.path, sample_id, mag_id)
+
+        else:
+            mag_id = None
+            for sample_id, file_paths in annotation_tuple:
+                copy_files(file_paths, partitioned_annotation.path, sample_id)
+
+        # Set key for partitioned_annotations dict to mag_id or sample_id
+        partitioned_annotation_key = mag_id if mag_id else sample_id
+
+        # Add the partitioned object to the collection dict
+        if num_partitions == num_annotations:
+            partitioned_annotations[partitioned_annotation_key] = partitioned_annotation
+        else:
+            partitioned_annotations[i] = partitioned_annotation
+
+    return partitioned_annotations
+
+
+def copy_files(file_paths: list, *dst_path_components):
```

Comment:
Hey @VinzentRisch, just noticed this now: maybe `utils.py` would be a better place for this function? It could be used in other places as well (potentially) so maybe you could move it there? Also, can you please add a test or two for this function?

Reply:
Yes will do, I forgot to ask you about the test because the function is covered by the other tests, so i wasn't sure if they are needed.

Reply:
Yup, as a rule of thumb, _always_ write tests for such functions as well (hence the name "unit" tests) - those will make it easier to discover bugs that may lurk in those small accessory functions 💪

Reply:
yeah makes total sense 👍